File: src/tools/visualStates/samples/trial10/trial10.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import sys, threading, time
import easyiceconfig as EasyIce
sys.path.append("/opt/jderobot/lib/python2.7")
sys.path.append("/opt/jderobot/share/jderobot/python3/visualStates_py")
from codegen.state import State
from codegen.temporaltransition import TemporalTransition
from codegen.conditionaltransition import ConditionalTransition
from codegen.python.runtimegui import RunTimeGui
from PyQt5.QtWidgets import QApplication


from jderobot import MotorsPrx

class State0(State):
	def runCode(self):
		pass


class State1(State):
	def runCode(self):
		pass


class State2(State):
	def runCode(self):
		interfaces->myMotors->setV(0.0);


class State3(State):
	def runCode(self):
		interfaces->myMotors->setV(0.2);


class State4(State):
	def runCode(self):
		interfaces->myMotors->setV(0.0);


class Tran1(TemporalTransition):

	def runCode(self):
		pass

class Tran5(TemporalTransition):

	def runCode(self):
		pass

class Tran2(TemporalTransition):

	def runCode(self):
		pass

class Tran4(TemporalTransition):

	def runCode(self):
		pass

class Interfaces():
	def __init__(self):
		self.ic = None
		self.myMotors = None
		self.connectProxies()

	def connectProxies(self):
		self.ic = EasyIce.initialize(sys.argv)
		self.myMotors = self.ic.propertyToProxy("automata.myMotors.Proxy")
		if not self.myMotors:
			raise Exception("could not create proxy with name:myMotors")
		self.myMotors = MotorsPrx.checkedCast(self.myMotors)
		if not self.myMotors:
			raise Exception("invalid proxy automata.myMotors.Proxy")

	def destroyProxies(self):
		if self.ic is not None:
			self.ic.destroy()

displayGui = False
guiThread = None
gui = None

def readArgs():
	global displayGui
	for arg in sys.argv:
		splitedArg = arg.split('=')
		if splitedArg[0] == '--displaygui':
			if splitedArg[1] == 'True' or splitedArg[1] == 'true':
				displayGui = True
				print('runtime gui enabled')
			else:
				displayGui = False
				print('runtime gui disabled')

def runGui():
	global gui
	app = QApplication(sys.argv)
	gui = RunTimeGui()
	gui.show()
	app.exec_()

if __name__ == "__main__":
	interfaces = Interfaces()

	readArgs()
	if displayGui:
		guiThread = threading.Thread(target=runGui)
		guiThread.start()


	if displayGui:
		while(gui is None):
			time.sleep(0.1)

		gui.addState(0, "root", True, 0.0, 0.0, None)
		gui.addState(1, "moving", True, 781.0, 764.0, 0)
		gui.addState(2, "stop", False, 944.0, 938.0, 0)
		gui.addState(3, "state 3", True, 817.0, 800.0, 1)
		gui.addState(4, "state 4", False, 984.0, 979.0, 1)

		gui.addTransition(1, "10secs", 1, 2, 803.5, 929.0)
		gui.addTransition(5, "transition 5", 2, 1, 903.5, 800.0)
		gui.addTransition(2, "transition 2", 3, 4, 865.5, 977.5)
		gui.addTransition(4, "transition 4", 4, 3, 969.5, 832.5)

	if displayGui:
		gui.emitActiveStateById(0)

	state0 = State0(0, True, interfaces, 100, None, gui)
	state1 = State1(1, True, interfaces, 100, state0, gui)
	state2 = State2(2, False, interfaces, 100, state0, gui)
	state3 = State3(3, True, interfaces, 100, state1, gui)
	state4 = State4(4, False, interfaces, 100, state1, gui)

	tran1 = Tran1(1, 2, 10000)
	state1.addTransition(tran1)

	tran5 = Tran5(5, 1, 5000)
	state2.addTransition(tran5)

	tran2 = Tran2(2, 4, 1000)
	state3.addTransition(tran2)

	tran4 = Tran4(4, 3, 1000)
	state4.addTransition(tran4)

	try:
		state0.startThread()
		state1.startThread()
		state0.join()
		state1.join()
		interfaces.destroyProxies()
		sys.exit(0)
	except:
		state0.stop()
		state1.stop()
		if displayGui:
			gui.close()
			guiThread.join()

		state0.join()
		state1.join()
		interfaces.destroyProxies()
		sys.exit(1)

File: src/tools/visualStates/samples/trial10/trial10.cpp
#include "trial10.h"
#include <iostream>
#include <string>
#include <runtimegui.h>

void State0::runCode() {
	
}

void State1::runCode() {
	
}

void State2::runCode() {
	interfaces->myMotors->setV(0.0);
}

void State3::runCode() {
	interfaces->myMotors->setV(0.2);
}

void State4::runCode() {
	interfaces->myMotors->setV(0.0);
}

void Tran1::runCode() {
	
}

void Tran5::runCode() {
	
}

void Tran2::runCode() {
	
}

void Tran4::runCode() {
	
}

void Interfaces::connectProxies(int argc, char* argv[]) {
	ice = EasyIce::initialize(argc, argv);

	Ice::ObjectPrx tempmyMotors = ice->propertyToProxy("automata.myMotors.Proxy");
	if (tempmyMotors == 0) {
		throw "cannot create proxy from automata.myMotors.Proxy";
	}
	myMotors = jderobot::MotorsPrx::checkedCast(tempmyMotors);
	if (myMotors == 0) {
		throw "invalid proxy automata.myMotors.Proxy";
	}
	std::cout << "myMotors is connected" << std::endl;
}

void Interfaces::destroyProxies() {
	if (ice != 0) {
		ice->destroy();
	}
}


pthread_t guiThread;
RunTimeGui* runTimeGui = NULL;
bool displayGui = false;

void readArgs(int *argc, char* argv[]) {
	int i;
	std::string splitedArg;

	for(i = 0; i < *argc; i++) {
		splitedArg = strtok(argv[i], "=");
		if (splitedArg.compare("--displaygui") == 0){
			splitedArg = strtok(NULL, "=");
			if (splitedArg.compare("true") == 0 || splitedArg.compare("True") == 0){
				displayGui = true;
				std::cout << "displayGui ENABLED" << std::endl;
			}else{
				displayGui = false;
				std::cout << "displayGui DISABLED" << std::endl;
			}
		}
		if(i == *argc -1){
			(*argc)--;
		}
	}
}

void* runGui(void*) {
	system("./trial10_runtime.py");
}

int main(int argc, char* argv[]) {
	Interfaces interfaces;
	try {
		interfaces.connectProxies(argc, argv);
	} catch (const Ice::Exception& ex) {
		std::cerr << ex << std::endl;
		interfaces.destroyProxies();
		return 1;
	} catch (const char* msg) {
		std::cerr << msg << std::endl;
		interfaces.destroyProxies();
		return 1;
	}

	readArgs(&argc, argv);

	if (displayGui) {
		pthread_create(&guiThread, NULL, &runGui, NULL);		runTimeGui = new RunTimeGui();

	}
	State* state0 = new State0(0, true, &interfaces, 100, NULL, runTimeGui);
	State* state1 = new State1(1, true, &interfaces, 100, state0, runTimeGui);
	State* state2 = new State2(2, false, &interfaces, 100, state0, runTimeGui);
	State* state3 = new State3(3, true, &interfaces, 100, state1, runTimeGui);
	State* state4 = new State4(4, false, &interfaces, 100, state1, runTimeGui);

	Transition* tran1 = new Tran1(1, 2, 10000);	state1->addTransition(tran1);
	Transition* tran5 = new Tran5(5, 1, 5000);	state2->addTransition(tran5);
	Transition* tran2 = new Tran2(2, 4, 1000);	state3->addTransition(tran2);
	Transition* tran4 = new Tran4(4, 3, 1000);	state4->addTransition(tran4);

	state0->startThread();
	state1->startThread();

	state0->join();
	state1->join();
}
